File: orchard/validators/typescript/src/validator.ts
/**
 * The validate() function — auto-selects JSON Schema draft based on the
 * schema's `$schema` field and runs Ajv, mapping each failure into our
 * normalized ValidationError shape.
 */
import { readFileSync } from 'node:fs';

import AjvDefault from 'ajv';
import Ajv2020Default from 'ajv/dist/2020.js';
import addFormatsDefault from 'ajv-formats';

import type { Result, ValidationError } from './errors.js';

// ESM-default interop — these packages still publish a `default` export under
// some moduleResolution settings. Coerce to the constructor either way.
const Ajv = ((AjvDefault as unknown as { default?: typeof AjvDefault }).default
  ?? AjvDefault) as typeof AjvDefault;
const Ajv2020 = ((Ajv2020Default as unknown as { default?: typeof Ajv2020Default }).default
  ?? Ajv2020Default) as typeof Ajv2020Default;
const addFormats = ((addFormatsDefault as unknown as { default?: typeof addFormatsDefault }).default
  ?? addFormatsDefault) as typeof addFormatsDefault;

const DRAFT_2020 = 'https://json-schema.org/draft/2020-12/schema';
const DRAFT_07_HASH = 'http://json-schema.org/draft-07/schema#';
const DRAFT_07_NO_HASH = 'http://json-schema.org/draft-07/schema';

type AnyAjv = InstanceType<typeof Ajv> | InstanceType<typeof Ajv2020>;

function selectAjv(schema: Record<string, unknown>): AnyAjv {
  const declared = typeof schema['$schema'] === 'string' ? (schema['$schema'] as string) : '';
  if (declared === DRAFT_2020) {
    const ajv = new Ajv2020({ strict: false, allErrors: true });
    addFormats(ajv as unknown as InstanceType<typeof Ajv>);
    return ajv;
  }
  if (declared === DRAFT_07_HASH || declared === DRAFT_07_NO_HASH || declared === '') {
    const ajv = new Ajv({ strict: false, allErrors: true });
    addFormats(ajv);
    return ajv;
  }
  // Unknown draft — fall back to default (draft-07 compatible)
  const ajv = new Ajv({ strict: false, allErrors: true });
  addFormats(ajv);
  return ajv;
}

function formatPath(instancePath: string | undefined): string {
  return instancePath && instancePath.length > 0 ? instancePath : '<root>';
}

function loadSchema(schema: object | string): Record<string, unknown> {
  if (typeof schema === 'string') {
    return JSON.parse(readFileSync(schema, 'utf-8'));
  }
  return schema as Record<string, unknown>;
}

/**
 * Validate `payload` against `schema`.
 *
 * @param schema A parsed JSON Schema object OR a filesystem path to a `.json` file.
 * @param payload Any JSON-compatible value.
 * @returns A `Result` with `ok` (boolean) and `errors` (readonly array).
 *
 * Throws if the schema cannot be compiled (catches malformed schemas early).
 */
export function validate(schema: object | string, payload: unknown): Result {
  const parsed = loadSchema(schema);
  const ajv = selectAjv(parsed);
  const fn = ajv.compile(parsed);
  const ok = fn(payload);

  if (ok) {
    return { ok: true, errors: [] };
  }

  const errors: ValidationError[] = (fn.errors ?? []).map((err): ValidationError => ({
    path: formatPath(err.instancePath),
    keyword: err.keyword,
    message: err.message ?? '',
    severity: 'error',
    params: {
      ajv_params: err.params as Record<string, unknown>,
      schema_path: err.schemaPath ?? '',
    },
  }));

  return { ok: false, errors };
}

File: orchard/validators/python/src/frootai_orchard/validator.py
"""The validate() function — auto-selects JSON Schema draft and runs
jsonschema's iter_errors, mapping each failure into our normalized
ValidationError shape.
"""
from __future__ import annotations

import json
from collections.abc import Iterable
from pathlib import Path
from typing import Any

from jsonschema import Draft7Validator, Draft202012Validator, FormatChecker
from jsonschema.validators import validator_for

from .errors import Result, ValidationError

# Map declared $schema URIs to validator classes.
_DRAFT_TO_VALIDATOR: dict[str, type] = {
    "https://json-schema.org/draft/2020-12/schema": Draft202012Validator,
    "http://json-schema.org/draft-07/schema#": Draft7Validator,
    "http://json-schema.org/draft-07/schema": Draft7Validator,
}


def _format_path(parts: Iterable[Any]) -> str:
    """Render a jsonschema absolute_path into a JSON Pointer-ish string."""
    parts = list(parts)
    if not parts:
        return "<root>"
    return "".join(f"/{p}" for p in parts)


def _select_validator_cls(schema: dict[str, Any]) -> type:
    declared = schema.get("$schema", "")
    if declared in _DRAFT_TO_VALIDATOR:
        return _DRAFT_TO_VALIDATOR[declared]
    # Fallback heuristic from jsonschema itself.
    return validator_for(schema)


def _coerce_param(value: Any) -> Any:
    """Make `validator_value` JSON-serializable. Drop callables / objects we
    can't safely round-trip; coerce sets/tuples to lists.
    """
    if value is None or isinstance(value, (bool, int, float, str)):
        return value
    if isinstance(value, (list, tuple, set)):
        return [_coerce_param(v) for v in value]
    if isinstance(value, dict):
        return {str(k): _coerce_param(v) for k, v in value.items()}
    return repr(value)


def validate(schema: dict[str, Any] | str | Path, payload: Any) -> Result:
    """Validate `payload` against `schema`.

    Args:
        schema: A parsed JSON Schema dict OR a path (str / Path) to a .json file.
        payload: Any JSON-compatible value.

    Returns:
        Result(ok=bool, errors=tuple[ValidationError, ...]).

    Raises:
        FileNotFoundError: if `schema` is a path and doesn't exist.
        json.JSONDecodeError: if the schema file isn't valid JSON.
        jsonschema.SchemaError: if the schema itself is malformed.
    """
    if isinstance(schema, (str, Path)):
        with open(schema, "r", encoding="utf-8") as f:
            schema = json.load(f)

    validator_cls = _select_validator_cls(schema)
    # Meta-validate first — catches schema bugs early.
    validator_cls.check_schema(schema)

    validator = validator_cls(schema, format_checker=FormatChecker())
    raw_errors = list(validator.iter_errors(payload))

    if not raw_errors:
        return Result(ok=True, errors=())

    errors = tuple(
        ValidationError(
            path=_format_path(err.absolute_path),
            keyword=err.validator or "unknown",
            message=err.message,
            severity="error",
            params={
                "validator_value": _coerce_param(err.validator_value),
                "schema_path": _format_path(err.absolute_schema_path),
            },
        )
        for err in raw_errors
    )
    return Result(ok=False, errors=errors)
